### PR TITLE
FreeBSD backend

### PR DESCRIPTION
**Pull request by arrowd**

### CODE REVIEW

**Comment by ximion** [comment]
> Some Linux distributions as well as FreeBSD OS install 3rd party software into
a prefix that differs from "/usr". This change makes asgen work in such cases.

Can you elaborate on that? Installing metadata into any location outside of `/usr/(local)/share/metainfo` is not permitted and not supported by any other tool, so I would really not like to make an exception here...

**Comment by ximion** [comment]
Also, by changing this you would not catch all system software that apparently is installed in `/usr`, *and* by making this a global build-time flag you can not process data for any other system other than FreeBSD correctly. So you can't use `appstream-generator` on FreeBSD to process a Debian or Arch Linux archive (or vice versa).

**Comment by arrowd** [comment]
> Can you elaborate on that?

FreeBSD installs 3rdpart software into `/usr/local`. Some Linuxes, AFAIK, install stuff into `/opt`.

> Also, by changing this you would not catch all system software that apparently is installed in /usr

I was under assumption that there could be no system software that have AppStream metadata. At least there are no in FreeBSD. But this point is valid.

> and by making this a global build-time flag you can not process data for any other system other than FreeBSD correctly. So you can't use appstream-generator on FreeBSD to process a Debian or Arch Linux archive (or vice versa).

I didn't even think about such use case. Yes, this is a valid point too.

I'm happy to fix this, but how? For FreeBSD backend to work I need asgen to use `/usr/local` prefix everywhere and without the first commit of this PR it is not possible.

**Comment by ximion** [comment]
> FreeBSD installs 3rdpart software into `/usr/local`. Some Linuxes, AFAIK, install stuff into `/opt`.

But do FreeBSD *packages* also contain stuff in `/usr/local`?
`/opt` is not a supported location at all for AppStream metadata, `/usr/local/share/metainfo` is fairly uncommon for packaged stuff, but that location is at least read, so we could potentially also support it.

**Comment by arrowd** [comment]
Bump. Can we get this in?

**Review by ximion — CHANGES_REQUESTED** [review]
Just a tiny nitpick about the TODO entry in the code, and the general disclaimer that I can't really test the code.
But apart from that, this looks good to go (I'd just like some feedback if the todo item is addressed, merging new code with that is a bit odd).

The patch also needs a (likely very simple) rebase on the master branch :-)

File: src/asgen/backends/freebsd/fbsdpkg.d [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    PackageKind _kind;
+
+    ArchiveDecompressor pkgArchive;
+    string[] contentsL = null;
+
+public:
+    this (string pkgRoot, JSONValue[string] j)
+    {
+        pkgjson = j;
+        pkgFname = buildPath (pkgRoot, pkgjson["path"].str()); // TODO: or pkgjson["repopath"] ???
```
[the remark this answers is not in the document]
Comment (arrowd) [reply src/asgen/backends/freebsd/fbsdpkg.d:48]: @bapt Can you help me here? I'm parsing packagesite lines in this code and using "path" entry to get "All/package-name-1.2.3.pkg" portion of the filename. The "repopath" entry contains the same value in the metadata from the official FreeBSD cluster. What's the difference and what's more correct to use there?

**Comment by arrowd** [comment]
> Just a tiny nitpick about the TODO entry in the code, and the general disclaimer that I can't really test the code.

We're running it for about a month now, so it works for us.

File: src/asgen/backends/freebsd/fbsdpkg.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (bapt) [reply src/asgen/backends/freebsd/fbsdpkg.d:48]: if I understand correctly what you are asking for, repopath is the path entry of the packagesite lines.

**Comment by ximion** [comment]
> We're running it for about a month now, so it works for us.

That's good enough - if the conflicts of this PR with the master branch are resolved, we can merge this (and maybe even remove the TODO, since it obviously works).

**Comment by arrowd** [comment]
This should be ready to get in now.

**Review by ximion — APPROVED** [review]

**Comment by ximion** [comment]
> This should be ready to get in now.

It is indeed! Thank you for your patch and your patience!

**Pull request merged**